pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -821,6 +821,19 @@ public void segmentStoppedConsuming(LLCSegmentName llcSegmentName, String instan
       _controllerMetrics.addMeteredTableValue(realtimeTableName, ControllerMeter.LLC_ZOOKEEPER_UPDATE_FAILURES, 1L);
       throw e;
     }
+    // We know that we have successfully set the idealstate to be OFFLINE.
+    // We can now do a best effort to reset the externalview to be OFFLINE if it is in ERROR state.
+    // If the externalview is not in error state, then this reset will be ignored by the helix participant
+    // in the server when it receives the ERROR to OFFLINE state transition.
+    // Helix throws an exception if we try to reset state of a partition that is NOT in ERROR state in EV,
+    // So, if any exceptions are thrown, ignore it here.
+    try {
+      _helixAdmin.resetPartition(_helixManager.getClusterName(), instanceName,
+          TableNameBuilder.REALTIME.tableNameWithType(llcSegmentName.getTableName()),
+          Collections.singletonList(segmentName));
+    } catch (Exception e) {
+      // Ignore
```

Comment:
So, as I described in the comments, this call from the server may come in one of two situations:
- The server could not start consumption at all (could not construct the consumer), so the state transition from OFFLINE to CONSUMING failed. The externalview will be set to ERROR by Helix. 
- The server succeeded the OFFLINE to CONSUMING state transition, but could not consume rows from the stream at some point after the state transition. There could be zero or any number of rows already consumed. In this case, the externalview is in CONSUMING state.

If we call the reset API in the second case, helix throws an exception. If we call in the first case, helix simply  does its best to reset the state and send an ERROR to OFFLINE state transition to the server.

Logging an error in this case will make the operator think that we did not reguister the offline call from server correctly. I am therefore not in favor of logging an error.

Reply:
@sajjad-moradi  and I discussed this, and we think the ideal solution is to:
- Create specific reason codes (not strings) that we can exchange between the controller and server when the server posts any event to the controller (not just to indicate stop consumption)
- Use the reason code to decide whether the reset API should be invoked (e.g. a reason code that indicates that the consuming segment was never created, will end up invoking the helix reset API). I will file an issue for this as a cleanup task.

Another alternative can be to look at the externalview in this same method before invoking the API. I had considered that alternative, and discarded it since it involves another Helix API call that may come in with its own errors and exceptions to be handled. It seemed to me to better to just call the reset API once and for all.